README.md
# Music-Generation
An NLP project to generate novel music

### Requirements

- `tensorflow` `2.4.0`
- `selenium` to scrape?
- `scipy` to signal process if needed

### Major TODOs

- Implement a data input pipeline (dataset needs to be found for this - leaning towards dataset from [Mutopia](https://www.mutopiaproject.org/). Might have to scrape the site for the download links)
- Implement 1D causal convolutions (PixelCNNs) and hence the entire wavenet model
- Experiment with model sizes and hyperparams 


### Future

- Vary architecture, search if others work better
- Extend to speech generation?


changelog.md
# Change Log

### 15-Sept-2021 : updated gan.py and input.py, Abhay
Finished up the gan specific functions, and most of the spectogram normalization and resizing functions (lifted straight from specgan). 
Implementing an inverse spectogram function is yet to be done. Debugging and optimization after that.

### 7-Sept-2021 : updated gan.py and input.py slightly, Abhay
Added a save function to the generator model every 20 epochs.
Also added a function in input.py to convert the audio arrays into melspectograms. Added a display function if we want to see the generated spectograms.
TODO: Discuss how to normalize (to (0, 1)) and denormalize the spectograms. Didn't really get how they did it in specgan

### 2-Sept-2021 : implemeted VAE model, Arvind
Added code to train.py to train a VAE model.
TODO: Figure out the input_shape.

### 23-Aug-2021 : added gan.py, Abhay
Added gan.py with the model definitions and the train function. SpecGAN architecture, with a couple differences. Used the basic loss function, can try out others later. 
TODO: A couple preprocessing and display functions, as well as a save function for the model at points. Will work on that.

###18-Aug-2021 : updated visualize_data in input.py, Arvind
Modified code in visualize_data to plot a graph for a given .wav file using librosa. TODO: Decide what to pass into conv layers.

### 16-Aug-2021 : updated input.py, Abhigyan###
Added code to the functions load_audio to input a filepath containing .wav files and return a list of tensors containing decoded float32 np arrays. Added functionality to play_sample using PyAudio that takes in a filename and plays the corresponding .wav audio. Made functional temporarily visualize_waveform using tensorflow(to be updated) and modified parameters of extract_data to include an option of instrument of choice. ToDo Add a readme file listing allowed inputs for the same.   

### 13-Aug-2021 : extract_data and (new)load_data_mono(incomplete), Abhay
Added code to 'extract_data' to store files of a certain instrument into a list, and shuffle them. Added a new function load_data_mono in order to work with mono music instead of stereo - a decision we will have to make. The code is still incomplete, have to decide on what exactly needs to be done - in what format Wavenet takes input. 
### 12-Aug-2021 : initial files, Balan
Added the `input.py` and `train.py` files with a rudimentary structure of functions to be implemented. More functions can be added if necessary. Defined classes for two models - VAE and WaveNet, to be implemented later. Download the IRMAS dataset from its mirror at [zenodo](https://zenodo.org/record/1290750#.YRQW1XUzbmx) (only the training set).

### {date} : {commit name}, {your name}
  {information about the commit, what\'s new how it affects other code (if it does)}


gan.py
import tensorflow as tf 
from tensorflow import keras 
from tensorflow.keras import layers
import os
import numpy as np
import random


def random_input(dimension, b_size):
    rand = np.reshape(np.random.randn(dimension * b_size), (b_size, dimension))
    
    return rand


def get_real(list, batch_size):
    index = random.sample(range(0, list.shape[0]), batch_size)
    X_real = list[index]
    y_real = np.ones((batch_size, 1))
    return X_real, y_real


def get_fake(model, size):
    generated = random_input(100, size)
    X_fake = model.predict(generated)
    y_fake = np.zeros((size, 1))


    return X_fake, y_fake

    
def generator():
    #Basically using SpecGAN with transposed convolutions instead of Conv + Upsampling layers, LeakyRELU instead of RELU, and Dropout
    #128 x 128 sized images are used, although the sampling rate used in SpecGAN is 16000. Can be modified
    model = keras.models.Sequential()
    model.add(layers.Dense(16384, input_dim = 100))
    model.add(layers.Reshape((4,4,1024)))
    model.add(layers.LeakyReLU(alpha = 0.2))
    model.add(layers.Conv2DTranspose(512, (5, 5), strides = (2, 2), padding = 'same'))
    model.add(layers.LeakyReLU(alpha = 0.2))
    model.add(layers.Dropout(0.4))
    model.add(layers.Conv2DTranspose(256, (5, 5), strides = (2, 2), padding = 'same'))
    model.add(layers.LeakyReLU(alpha = 0.2))
    model.add(layers.Dropout(0.4))
    model.add(layers.Conv2DTranspose(128, (5, 5), strides = (2, 2), padding = 'same'))
    model.add(layers.LeakyReLU(alpha = 0.2))
    model.add(layers.Dropout(0.4))
    model.add(layers.Conv2DTranspose(64, (5, 5), strides = (2, 2), padding = 'same'))
    model.add(layers.LeakyReLU(alpha = 0.2))
    model.add(layers.Dropout(0.4))
    model.add(layers.Conv2DTranspose(1, (5, 5), strides = (2, 2), padding= 'same'))
    model.add(layers.Activation('tanh'))

    return model

def discriminator():
    model = keras.models.Sequential()
    model.add(layers.Conv2D(64, (5, 5), strides = (2, 2), padding= 'same', input_shape = (128, 128, 1)))
    model.add(layers.LeakyReLU(alpha = 0.2))
    model.add(layers.Conv2D(128, (5, 5), strides = (2, 2), padding = 'same', kernel_initializer='he normal'))
    model.add(layers.LeakyReLU(alpha = 0.2))
    model.add(layers.Conv2D(256, (5, 5), strides = (2, 2), padding=  'same', kernel_initializer='he normal'))
    model.add(layers.LeakyReLU(alpha = 0.2))
    model.add(layers.Conv2D(512, (5, 5), strides = (2, 2), padding = 'same', kernel_initializer='he normal'))
    model.add(layers.LeakyReLU(alpha = 0.2))
    model.add(layers.Conv2D(1024, (5, 5), strides = (2, 2), padding= 'same', kernel_initializer='he normal'))
    model.add(layers.LeakyReLU(alpha = 0.2))
    model.add(layers.Flatten())
    model.add(layers.Dense(1,activation = 'sigmoid'))
    opt = keras.optimizers.Adam(learning_rate=0.0002, beta_1=0.5)
    model.compile(loss='binary crossentropy', optimizer=opt,metrics=['accuracy'])
    

    return model


def combined(gen,disc):
    disc.trainable = False
    model = keras.models.Sequential()
    model.add(gen)
    model.add(disc)
    opt = keras.optimizers.Adam(learning_rate=0.0002, beta_1=0.5)
    model.compile(loss= 'binary crossentropy', optimizer=opt)


    return model


def train_gan(model_gen, model_disc, directory, model_combined, epochs = 500, batch_size = 128): #To be tuned
    total_files = len([name for name in os.listdir('directory')])
    batch_number = int(total_files / batch_size)
    for i in range(epochs):
        for j in range(batch_number):
            Xr, yr = get_real(directory, batch_size)
            Xf, yf = get_fake(model_gen, batch_size)
            disc_loss_real, _ = model_disc.train_on_batch(Xr, yr)
            disc_loss_fake, _ = model_disc.train_on_batch(Xf, yf)
            disc_loss = 0.5 * (np.add(np.array(disc_loss_real), np.array(disc_loss_fake)))
            #Improvement: Try to implement Wasserstein Loss instead of this
            Xt = random_input(100, batch_size)
            yt = np.ones((batch_size, 1))
            gen_loss = model_combined.train_on_batch(Xt, yt)
            print('Epoch : %d/%d, Batch : %d/%d, DLoss: %.3f, GLoss : %.3f' % (i + 1, j + 1, disc_loss, gen_loss))
            
        if (i + 1) % 20 == 0:
            name = 'Epoch %d Model.h5' % (i + 1)
            model_gen.save(name)

    return

input.py
import matplotlib.pyplot as plt
import tensorflow as tf
import pyaudio
import os
import librosa, librosa.display
import random
import numpy as np



def extract_data(instr):
    # go through the files and pick out the right directory to use, shuffling, all that messing around
    path = 'G:\My Drive\Music Generation\IRMAS-TrainingData\\'+instr
    #Using a placeholder path that consists of (mainly) a single instrument to avoid the complexity of polyphonic music. Will have to be decided upon later.
    ### Add a function parameter asking which instrument to use, you can also provide a list of allowed strings in the readme or sth
    file_names = []
    for file in os.listdir(path):
        file_names.append(path + '\\' + file)
    random.shuffle(file_names)
    #Technically redundant, since the file names are appended in an arbitrary order
    return file_names
    #TODO:    Decide whether we want to use a single instrument. 

def load_audio(directory_list):
    # use decode_wav from the audio submodule to read *.wav files
    # get the list of files and read them into a tensor
    # split into train, val, test (is there a need since the dataset already has multiple testing sets)
    waveforms=list()
    for path in directory_list:
        parts=tf.strings.split(path, os.path.sep)
        label=parts[-2]                  #Are labels needed? Since we will be shuffling data randomly.
        audio_binary = tf.io.read_file(path)
        waveform, sr = tf.audio.decode_wav(audio_binary)
        waveforms.append(waveform)
        
    return waveforms
    

def play_sample(filename):
    # pyaudio seems to be better than playsound (which we discussed because its not been maintained in a while)
    # given a file name make it play the sample
    chunk_size = 1024
    wf = wave.open(filename,'rb')
    interface = pyaudio.PyAudio()
    stream = interface.open(format=interface.get_format_from_width(wf.getsampwidth()), channels=wf.getnchannels(), rate=wf.getframerate(), output=True)
    
    aud_data = wf.readframes(chunk_size)
    
    while len(aud_data)>0:
        stream.write(aud_data)
        aud_data = wf.readframes(chunk_size)
    
    stream.stop_stream()
    stream.close()
    wf.close()
    interface.terminate()
    

def visualize_waveform(wave):
   #plot a graph for a given .wav file
    x,sr=librosa.load(wave)
    librosa.display.waveplot(x, sr)
    
def load_audio_mono(file_list):
    # In case we are using mono music (1 channel) instead of stereo  (2 channels), since it is easier, and the vector size is halved.
    ### use mono for now, if needed, we can implement stereo as well
    tensor_list = []
    for song in file_list:
        x, sr = librosa.load(song, sr=22050) 
        tensor_list.append[x]
    # TODO: Split each tensor into smaller samples? The audio clips themselves are only 3 seconds long, so there is barely any temporal dependency.
    #       Split into train,val,test - if using a certain instrument(?) it can be done manually, otherwise train_test_split twice.
    #       Make use of tf.data.Datasets for convenience in batching and the like.
    #       Figure out what exactly we are passing to the conv layers and in what shape.       
    pass


def audio_to_mel(list):
    mel_spec_list = []
    for sample in list:
        spect = librosa.feature.melspectogram(y = sample)
        mel_spec_list.append(spect)
    return mel_spec_list

def get_mean_and_std(spec_list):
    means = []
    stds = []
    for spec in spec_list:
        means.append(np.mean(spec, axis = 1))
        stds.append(np.std(spec, axis = 1))
    mean = np.mean(means, axis = 0)
    std = np.mean(stds, axis = 0)
    return mean, std

def normalize_mel(spectogram_list, spec_mean, spec_std):
    norm_spec = []
    for item in spectogram_list:
        norm_item = (item - spec_mean) / (3.0 * spec_std)  
        clipped = np.clip(norm_item, -1.0, 1.0)
        norm_spec.append(clipped)
    return norm_spec


def denormalize_mel(melspectogram_list, mean_spec, std_spec):
    denorm_spec = []
    for gram in melspectogram_list:
        denorm_gram = (gram * (3.0 * std_spec)) + mean_spec
        denorm_spec.append(denorm_gram)
    return denorm_spec

def load_spec(mel_list):
    append = np.ones((128, 128)) * (-80)
    right_size_mel = []
    for melspec in mel_list:
        melspec = np.hstack((melspec,append))
        melspec = melspec[:, :128]
        right_size_mel.append(melspec)
    return right_size_mel
        


def mel_to_audio(spec):
    pass


def plot_spectograms(melspec_list):
    #If we want to see the denormalized spectograms for some reason. Here 9 will be shown. Changeable.
    for i in range(9):
        plt.subplot(3, 3, i + 1)
        plt.axis('off')
        plt.imshow(melspec_list[i])
    plt.show()


pixelcnn.py
#!/usr/bin/env python
# coding: utf-8

# In[31]:


import tensorflow as tf
from tensorflow import keras
from tensorflow.keras import layers
import numpy as np


# In[32]:


from keras.utils.generic_utils import get_custom_objects
from keras.layers import Activation
from keras.backend import sigmoid


# In[33]:


def activation_funtion(x):  #creating a gated activation function for use
    return np.dot(tanh(x),sigmoid(x))

get_custom_objects().update({'gated': Activation(activation_funtion)})


# In[35]:


#PixelCNN Layer
class ConvLayer(layers.Layer,):
    def __init__(self):
        super(PixelConvLayer, self).__init__()
        self.mask_type = mask_type
        self.convolution = layers.Conv2D(**kwargs)
        
    
    def Create(self,input_shape):
        self.convolution.Create(input_shape)
        kernel_shape = self.convolution.kernel.get_shape()
        self.maks = np.zeros(shape=kernel_shape)
        self.mask[: kernel_shape[0] // 2, ...] = 1.0
        self.mask[kernel_shape[0] // 2, : kernel_shape[1] // 2, ...] = 1.0
        if self.mask_type == "B":
            self.mask[kernel_shape[0] // 2, kernel_shape[1] // 2, ...] = 1.0
        
    
    def Call(self):
        self.convolution.kernel.assign(self.convolution.kernel * self.mask)
        return self.convolution(inputs)


# In[36]:


#Residual Skip Connection layer
class ResidualSkipConnect(layers.Layer):
    def __init__(self,filters,**kwargs):
        super(ResidualSkipConnect,self).__init__(**kwargs)
        self.convolution1 = keras.layers.Conv2D(filters=filters,kernel_size=1,activation='gated' )
        self.pixel_convolution = ConvLayer(mask_type='B',filters=filters//2,kernel_size=2,activation='gated',padding='same')
        self.convolution2 = keras.layers.Conv2D(filters=filters,kernel_size=1,activation='gated')
        
    def Call(self,input):
        pass 


# In[28]:


#Parameterized Skip Connection Layer
class ParamSkipConnect():
    def __init__(self):
        pass


# In[ ]:






train.py
from tensorflow.keras import layers
from tensorflow.keras import models
import tensorflow as tf
import tensorflow_addons as tfa
import time
import numpy as np
from IPython import display
from IPython.display import clear_output


array_shape=9000
'''input_shape not sure about this'''
latent_dim=2
epochs=20
'''need to figure out the values to use'''

class Resnet1DBlock(tf.keras.Model):
    def __init__(self, kernel_size, filters,type='encode'):
        super(Resnet1DBlock, self).__init__(name='')
    
        if type=='encode':
            self.conv1a = layers.Conv1D(filters, kernel_size, 2,padding="same")
            self.conv1b = layers.Conv1D(filters, kernel_size, 1,padding="same")
            self.norm1a = tfa.layers.InstanceNormalization()
            self.norm1b = tfa.layers.InstanceNormalization()
        if type=='decode':
            self.conv1a = layers.Conv1DTranspose(filters, kernel_size, 1,padding="same")
            self.conv1b = layers.Conv1DTranspose(filters, kernel_size, 1,padding="same")
            self.norm1a = tf.keras.layers.BatchNormalization()
            self.norm1b = tf.keras.layers.BatchNormalization()
        else:
            return None

    def call(self, input_tensor):
        x = tf.nn.relu(input_tensor)
        x = self.conv1a(x)
        x = self.norm1a(x)
        x = layers.LeakyReLU(0.4)(x)

        x = self.conv1b(x)
        x = self.norm1b(x)
        x = layers.LeakyReLU(0.4)(x)

        x += input_tensor
        return tf.nn.relu(x)

class VAE_model(tf.keras.Model):
    def __init__(self, latent_dim):
        super(VAE_model, self).__init__()
        self.latent_dim = latent_dim
        self.encoder = tf.keras.Sequential(
            [
                tf.keras.layers.InputLayer(input_shape=(1,array_shape)),
                layers.Conv1D(64,1,2),
                Resnet1DBlock(64,1),
                layers.Conv1D(128,1,2),
                Resnet1DBlock(128,1),
                layers.Conv1D(128,1,2),
                Resnet1DBlock(128,1),
                layers.Conv1D(256,1,2),
                Resnet1DBlock(256,1),
                # No activation
                layers.Flatten(),
                layers.Dense(latent_dim+latent_dim)

            ]
        )
        self.decoder = tf.keras.Sequential(
            [
                tf.keras.layers.InputLayer(input_shape=(latent_dim,)),
                layers.Reshape(target_shape=(1,latent_dim)),
                Resnet1DBlock(512,1,'decode'),
                layers.Conv1DTranspose(512,1,1),
                Resnet1DBlock(256,1,'decode'),
                layers.Conv1DTranspose(256,1,1),
                Resnet1DBlock(128,1,'decode'),
                layers.Conv1DTranspose(128,1,1),
                Resnet1DBlock(64,1,'decode'),
                layers.Conv1DTranspose(64,1,1),
                # No activation
                layers.Conv1DTranspose(array_shape,1,1),
            ]
        )

    def sample(self, eps=None):
        if eps is None:
            eps = tf.random.normal(shape=(200, self.latent_dim))
        return self.decode(eps, apply_sigmoid=True)
    
    def encode(self, x):
        mean, logvar = tf.split(self.encoder(x), num_or_size_splits=2, axis=1)
        return mean, logvar

    def reparameterize(self, mean, logvar):
        eps = tf.random.normal(shape=mean.shape)
        return eps * tf.exp(logvar * .5) + mean

    def decode(self, z, apply_sigmoid=False):
        logits = self.decoder(z)
        if apply_sigmoid:
            probs = tf.sigmoid(logits)
            return probs
        return logits

        
model = VAE_model(latent_dim)
optimizer = tf.keras.optimizers.Adam(0.0003,beta_1=0.9, beta_2=0.999,epsilon=1e-08)
def log_normal_pdf(sample, mean, logvar, raxis=1):
    log2pi = tf.math.log(2. * np.pi)
    return tf.reduce_sum(
         -.5 * ((sample - mean) ** 2. * tf.exp(-logvar) + logvar + log2pi),
          axis=raxis)
def compute_loss(model, x):
    mean, logvar = model.encode(x)
    z = model.reparameterize(mean, logvar)
    x_logit = model.decode(z)
    cross_ent = tf.nn.sigmoid_cross_entropy_with_logits(logits=x_logit, labels=x)
    logpx_z = -tf.reduce_sum(cross_ent, axis=[1,2])
    logpz = log_normal_pdf(z, 0., 0.)
    logqz_x = log_normal_pdf(z, mean, logvar)
    return -tf.reduce_mean(logpx_z + logpz - logqz_x)

def train_step(model, x, optimizer): 
    """Executes one training step and returns the loss.
       This function computes the loss and gradients, and uses the latter to
       update the model's parameters.
     """
    with tf.GradientTape() as tape:
            loss_KL = compute_loss(model,x)
            mean, logvar = model.encode(x)
            z = model.reparameterize(mean, logvar)
            x_logit = model.decode(z)
            reconstruction_loss = tf.reduce_mean(tf.keras.losses.binary_crossentropy(x, x_logit))
            total_loss = reconstruction_loss+ loss_KL
    gradients = tape.gradient(total_loss, model.trainable_variables)
    optimizer.apply_gradients(zip(gradients, model.trainable_variables))

def train_VAE(epoch,train_dataset,test_dataset,model,optimizer):
  for epoch in range(1, epochs + 1):
    start_time = time.time()
    for train_x in train_dataset:
        train_x = np.asarray(train_x)[0]
        train_step(model, train_x, optimizer)
    end_time = time.time()

    loss = tf.keras.metrics.Mean()
    for test_x in test_dataset:
        test_x = np.asarray(test_x)[0]
        loss(compute_loss(model, test_x))
    display.clear_output(wait=False)
    elbo = -loss.result()
    print('Epoch: {}, Test set ELBO: {}, time elapse for current epoch: {}'.format(epoch, elbo, end_time - start_time))



class wavenet_model:
    def __init__(self, inputs, outputs):
        # define some params
        pass

    def forward(self):
        # run a forward pass
        pass

    def optimize(self):
        # calculate loss and add an optimizer
        pass

    def accuracy(self):
        # calculate accuracy
        pass

